source/module_elecstate/module_charge/charge_extra.cpp

```diff
@@ -93,60 +103,59 @@ void Charge_Extra::extrapolate_charge(
 
     // if(lsda || noncolin) rho2zeta();
 
-    // read charge difference into chr->rho
-    read_files(
-#ifdef __MPI
-        Pgrid,
+    double** rho_atom = new double*[GlobalV::NSPIN];
+    for (int is = 0; is < GlobalV::NSPIN; is++)
+    {
+        rho_atom[is] = new double[chr->rhopw->nrxx];
+    }
+    chr->atomic_rho(GlobalV::NSPIN, omega_old, rho_atom, sf->strucFac, ucell);
+#ifdef _OPENMP
+#pragma omp parallel for collapse(2) schedule(static, 512)
 #endif
-        chr->rhopw->nx,
-        chr->rhopw->ny,
-        chr->rhopw->nz,
-        &ucell,
-        "NOW",
-        chr->rho);
+    for (int is = 0; is < GlobalV::NSPIN; is++)
+    {
+        for (int ir = 0; ir < chr->rhopw->nrxx; ir++)
+        {
+            chr->rho[is][ir] -= rho_atom[is][ir];
+            chr->rho[is][ir] *= omega_old;
+        }
+    }
 
     if(rho_extr == 1)
     {
         GlobalV::ofs_running << " NEW-OLD atomic charge density approx. for the potential !" << std::endl;
+
+        if (pot_order > 1)
+        {
+#ifdef _OPENMP
+#pragma omp parallel for collapse(2) schedule(static, 512)
+#endif
+            for (int is = 0; is < GlobalV::NSPIN; is++)
+            {
+                for (int ir = 0; ir < chr->rhopw->nrxx; ir++)
+                {
+                    delta_rho1[is][ir] = chr->rho[is][ir];
+                }
+            }
+        }
     }
     // first order extrapolation
     else if(rho_extr ==2)
     {
         GlobalV::ofs_running << " first order charge density extrapolation !" << std::endl;
 
-        // read charge difference into delta_rho1
-        double** delta_rho1 = new double*[GlobalV::NSPIN];
-        for (int is = 0; is < GlobalV::NSPIN; is++)
-        {
-            delta_rho1[is] = new double[chr->rhopw->nrxx];
-        }
-        read_files(
-#ifdef __MPI
-            Pgrid,
-#endif
-            chr->rhopw->nx,
-            chr->rhopw->ny,
-            chr->rhopw->nz,
-            &ucell,
-            "OLD1",
-            delta_rho1);
-
 #ifdef _OPENMP
 #pragma omp parallel for collapse(2) schedule(static, 128)
 #endif
-        for(int is=0; is<GlobalV::NSPIN; is++)
+        for (int is = 0; is < GlobalV::NSPIN; is++)
         {
-            for(int ir=0; ir<chr->rhopw->nrxx; ir++)
+            for (int ir = 0; ir < chr->rhopw->nrxx; ir++)
```

Comment:
rhopw->nrxx can be replaced by nrxx (the input parameter)?

Reply:
This function will call functions of class Charge, so I think it is unnecessary to add an extra parameter here.